packages/i18next/src/lib/functions.ts

```diff
@@ -0,0 +1,188 @@
+import { container } from '@sapphire/pieces';
+import { isObject, NonNullObject } from '@sapphire/utilities';
+import { Message, MessageOptions } from 'discord.js';
```

Comment:
Agnosticity was always a pain that made the types a lot harder than they should have been.

Besides, there's no real need to use an alternative Discord library for Node.js starting with discord.js v13.

Reply:
Maybe for you, but for some of us (hi) we would like to use other things than discord.js, tyvm. Maybe not a priority, but agnosticity should be something we ought to achieve

Reply:
> Agnosticity was always a pain that made the types a lot harder than they should have been.

Not necessarily. We demonstrated the ability to use general abstractions to make extensible registers, dynamic class mixins, et cetera. A pain, perhaps, but we managed it - and it worked well.

> Besides, there's no real need to use an alternative Discord library for Node.js starting with discord.js v13.

That's a relatively close minded perspective. What if one does come along, and it turns out to be better than discord.js? Or what if one does come along, and it's different in a fundamental way that works better for some use cases?

As was our philosophy, it is not our place to dictate what people use. We are a framework, not a library - and a framework's job is to help people build things, not to mandate how they build them.

Reply:
> Not necessarily. We demonstrated the ability to use general abstractions to make extensible registers, dynamic class mixins, et cetera. A pain, perhaps, but we managed it - and it worked well.

Except it relied on type hacks and made documentation basically `null` because there's no way to properly document self-augmentation with the current types.

Go complain TypeScript for not making something like C/C++/C#/Rust's preprocessor conditions, I face this pain even in canvas-constructor.

> That's a relatively close minded perspective. What if one does come along, and it turns out to be better than discord.js? Or what if one does come along, and it's different in a fundamental way that works better for some use cases?

We'll look into that when there's a native way for TypeScript to make this work without messy and hacky workarounds.

Reply:
> Except it relied on type hacks and made documentation basically `null` because there's no way to properly document self-augmentation with the current types.

I have no idea if you've ever seen the documentation for it but it isn't null at all because it's built out of a series of interfaces. Everything is typed properly and documented properly - the only thing with no documentation is how those interfaces are hooked up to discord.js' types.

> Go complain TypeScript for not making something like C/C++/C#/Rust's preprocessor conditions, I face this pain even in canvas-constructor.

It'd take a while, but I'm open to drafting an RFC to TypeScript for intentionally dynamic type augmentation? Up to you.

Reply:
In fact, I've just noticed something. It isn't particularly hacky at all - it does a relatively standard thing by declaring the module and re-exporting as an extension. In fact, that's exactly what Sapphire does to augment discord.js.

All the register does is augment the base package with its own extensions, and customize the implementation for each function which is laid out in the general interfaces.

These are all standard features of TypeScript, right the way from dynamic class mixins to module augmentation and interface implementation. The whole thing. So how is it hacky? If anything, it's a marvel of engineering, and a testament to the things you can achieve with TypeScript's inner workings.